src/reducers/tabs.js

```diff
@@ -0,0 +1,107 @@
+// @flow
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Tabs reducer
+ * @module reducers/tabs
+ */
+
+import * as I from "immutable";
+import makeRecord from "../utils/makeRecord";
+import { prefs } from "../utils/prefs";
+
+import type { List } from "immutable";
+import type { Action } from "../actions/types";
+import type { Record } from "../utils/makeRecord";
+
+type Tab = {
+  id: string,
+  title: string,
+  tooltip?: string
+};
+
+type TabList = List<Tab>;
+export type TabsState = {
+  currentTabIndex: number
+  tabs: TabList
+}
+
+export function initialState(): Record<TabsState> {
+  return makeRecord({
+    currentTabIndex: 0,
+    tabs: I.List()
+  })
+}
+
+export default function update(state: Record<TabsState>, action: Action): Record<TabsState> {
+  switch (action.type) {
+    case "ADD_TAB":
+      const tab = { id: action.id, title: action.title, tooltip: action.tooltip };
+      return state.merge({ tabs: updateTabList({ tabsState: state }, tab, action.tabIndex) });
+
+    case "SELECT_TAB":
+      return state.set("currentTabIndex", action.tabIndex);
+
+    case "MOVE_TAB":
+      const tab = state.tabs.find(tab => tab.id === id);
+      return state.merge({
+        tabs: updateTabList({ tabsState: state }, tab, action.tabIndex)
+      });
+
+    case "CLOSE_TAB":
+      const tabs = removeFromTabList({ tabsState: state }, [action.id]);
+      prefs.tabs = tabs;
+      return state.merge({ tabs });
+
+    case "CLOSE_TABS":
+      const tabs = removeFromTabList({ tabsState: state }, action.ids);
+      prefs.tabs = tabs;
+      return state.merge({ tabs });
+  }
+
+  return state;
+};
+
+/**
+ * Adds the new tab to the list or moves the tab in the list if it is not already there
+ * @memberof reducers/sources
+ * @static
+ */
+function updateTabList(state: OuterState, currentTab: Tab, moveIndex?: number) {
```

Comment:
this function is doing two things, moving and adding -- lets split this into two functions.

Reply:
also, i investigated when we move tabs... it is rare. is there any information on why we have so much infrastructure for moving things?

Reply:
I think it was done initially, in the bid to cover all the possible moving scenarios, now we can make more informed improvements!